overlord/configstate/configcore/debug.go

```diff
@@ -0,0 +1,150 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
+
+/*
+ * Copyright (C) 2024 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package configcore
+
+import (
+	"fmt"
+	"os"
+	"path/filepath"
+
+	"github.com/snapcore/snapd/dirs"
+	"github.com/snapcore/snapd/osutil"
+	"github.com/snapcore/snapd/sysconfig"
+	"github.com/snapcore/snapd/systemd"
+)
+
+const (
+	optionDebugSnapdLog            = "debug.snapd.log"
+	optionDebugSystemdLogLevel     = "debug.systemd.log-level"
+	coreOptionDebugSnapdLog        = "core." + optionDebugSnapdLog
+	coreOptionDebugSystemdLogLevel = "core." + optionDebugSystemdLogLevel
+)
+
+func init() {
+	supportedConfigurations[coreOptionDebugSnapdLog] = true
+	supportedConfigurations[coreOptionDebugSystemdLogLevel] = true
+}
+
+func validateDebugSnapdLogSetting(tr ConfGetter) error {
+	return validateBoolFlag(tr, optionDebugSnapdLog)
+}
+
+func handleDebugSnapdLogConfiguration(_ sysconfig.Device, tr ConfGetter, opts *fsOnlyContext) error {
+	debugLog, err := coreCfg(tr, optionDebugSnapdLog)
+	if err != nil {
+		return err
+	}
+
+	rootDir := dirs.GlobalRootDir
+	if opts != nil {
+		rootDir = opts.RootDir
+	}
+	envDir := filepath.Join(dirs.SnapdStateDir(rootDir), "environment")
+
+	snapdEnvPath := filepath.Join(envDir, "snapd.conf")
+
+	switch debugLog {
+	case "true":
+		if err := os.Mkdir(envDir, 0755); err != nil && !os.IsExist(err) {
+			return err
+		}
+		if err := osutil.EnsureFileState(snapdEnvPath, &osutil.MemoryFileState{
+			Content: []byte("SNAPD_DEBUG=1\n"),
+			Mode:    os.FileMode(0644),
+		}); err != nil {
+			return err
+		}
+	case "false", "":
+		// We simply remove the env file as for the moment we use it
+		// just for SNAPD_DEBUG. If we change that we will need to
+		// locate the variable in the file and remove just that.
+		if err := os.Remove(snapdEnvPath); err != nil && !os.IsNotExist(err) {
+			return err
+		}
+	default:
+		return fmt.Errorf("%s must be true of false, not: %q", optionDebugSnapdLog, debugLog)
+	}
+
+	// TODO a snapd restart so the option is immediately applied. Change
```

Comment:
I wonder if we could just call `logger.SetLogger` instead. Or add a function that that changes "debug".

Reply:
Very good point, we don't really need to restart snapd indeed. What you suggest would involve some changes in the logger, I can do that as a follow-up. I've changed the TODO now.